pyro/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,144 @@
+        self.normalized = normalized
+        self.batch_size = batch_size if batch_size is not None else 1
+        if scale_tril is not None and covariance_matrix is not None:
+            raise ValueError("Only either 'covariance_matrix' or 'scale_tril' can be passed at once.")
+        if covariance_matrix is not None:
+            if not covariance_matrix.dim() == 2:
+                raise ValueError("The covariance matrix must be a matrix, but got covariance_matrix.size() = {}".format(
+                                    loc.size()))
+            self.sigma = covariance_matrix
+            self.sigma_cholesky = potrf_compat(covariance_matrix)
+        else:
+            if not scale_tril.dim() == 2:
+                raise ValueError("The Cholesky decomposition of the covariance matrix must be a matrix, "
+                                 "but got scale_tril.size() = {}".format(
+                                    loc.size()))
+            self.sigma = torch.mm(scale_tril.transpose(0, 1), scale_tril)
+            self.sigma_cholesky = scale_tril
+        if loc.dim() > 1:
+            raise ValueError("The mean must be a vector, but got loc.size() = {}".format(loc.size()))
+
+        super(MultivariateNormal, self).__init__(*args, **kwargs)
+
+    def batch_shape(self, x=None):
+        mu = self.mu.expand(self.batch_size, *self.mu.size()).squeeze(0)
+        if x is not None:
+            if x.size()[-1] != mu.size()[-1]:
+                raise ValueError("The event size for the data and distribution parameters must match.\n"
+                                 "Expected x.size()[-1] == self.mu.size()[0], but got {} vs {}".format(
+                                    x.size(-1), mu.size(-1)))
+            try:
+                mu = mu.expand_as(x)
+            except RuntimeError as e:
+                raise ValueError("Parameter `mu` with shape {} is not broadcastable to "
+                                 "the data shape {}. \nError: {}".format(mu.size(), x.size(), str(e)))
+
+        return mu.size()[:-1]
+
+    def event_shape(self):
+        return self.mu.size()[-1:]
+
+    def sample(self):
+        """
+        Generate a sample with the specified covariance matrix and mean.
+        Differentiation wrt. to the covariance matrix is only supported on PyTorch version 0.3.0 or higher.
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.sample`
+
+        """
+        batch_size = self.batch_size
+        uncorrelated_standard_sample = Variable(torch.randn(batch_size, *self.mu.size()).type_as(self.mu.data))
+        transformed_sample = self.mu + torch.mm(uncorrelated_standard_sample, self.sigma_cholesky)
+        return transformed_sample if self.reparameterized else transformed_sample.detach()
+
+    def batch_log_pdf(self, x):
+        if not self.normalized and self.sigma_cholesky.requires_grad:
+            warnings.warn("Gradients will not take normalization into account if normalized=False.")
```

Comment:
I'm inclined to raise an error rather than warn because I do not understand what is being computed in this case. Do you have a clear understanding of what is being computed and whether there are any statistical applications for computing this new quantity?

Reply:
Of course the gradients would only be a kind of rather bad estimate of the true gradient, but it looked to me like it was good enough to do some things like fitting the parameters of a kernel for a gaussian process. But I did investigate that very thoroughly so I will change the code to raise an error for the moment. It looks like (2) can be implemented very soon anyway.

Reply:
Thanks, raising an error seems like the right thing to do, and you've convinced me that a correct solution should be fairly cheap.